### ISSUE TITLE
Additional Label in TextMessageCell

### ISSUE THREAD

**Issue by LisforLyance** [badoo/Chatto#390]
Hey, 
I want to add an additional label below every message in addition to the avatar (see picture) to make everything a bit clearer in a group chat. Is ist possible to achieve this somehow by using the existing message cells, or should I just create a completely new element?

Sketch of the intended design:
https://imgur.com/a/USpU2

**Comment by ocastilloApplaudo** [comment]
I hope they answer this question.   I had to create a new collectionViewCell,  but the problem I've had is the height of the cell.  Height for cell is  called BEFORE   **func dequeueCell**   so I've had to calculate the height of my label and my other label programatically,  PLUS  the vertical constraints.   Storyboard constraints does nothing for the height  :(

**Comment by AntonPalich** [comment]
The easiest way to do it without modifying existing code is to provide an author as an additional item.

First, you have to build a new item that will display author. You can check the example here

**Comment by AntonPalich** [comment]
The easiest way to add authors is to add them as new items. This will allow you to reuse existing text cells. We do it in Badoo for message statuses (Sending..., Delivered, Read, Failed). We mix messages  with all other additional items (statuses, separators, loading indicator).

@LisforLyance, @ocastilloApplaudo I will write a tutorial about adding new items that behaves like messages (that support swipe to reveal frature, for example) in Chatto today/tomorrow. Until then, you can check this simple tutorial: https://github.com/badoo/Chatto/wiki/Tutorial_Add_Custom_Cell

**Comment by saroar** [comment]
@LisforLyance did you done it if yes would you like to share with me please i like to show each username and avatar Thanks advance

**Comment by LisforLyance** [comment]
@saroar Yes I did it just as @AntonPalich explained it. I created a **LabelCollectionViewCell**, a **LabelPresenter** with a **LabelPresenterBuilder** and a **LabelModel**. You can copy the classes of your **TimeSeparator** and just adjust them, so that they match your needs. 
You can then add the **LabelPresenterBuilder** inside your ChatViewControllers **createPresenterBuilders** method. You also have to create an instance of **LabelModel** in your Decorator class. Just add it inside of the **decorateItems** method. That is where you can also check, if the message is incoming or outgoing and then only display it for incoming.

I hope you can understand it. If you have trouble with it, just tell me and I can provide some sample code.

**Comment by saroar** [comment]
@LisforLyance i am very young in swift language only 3 months so I will be happy to see your code and screen short how your app u show it, if it is possible you can send email saroar9@gmail.com or create full gist and share with me, I already create demo app with 1to1 chat but for me group chat I need to show each your photo or name here my demo app https://github.com/saroar/RealtTimeIOSChat

**Comment by LisforLyance** [comment]
@saroar I am also new to Swift and programming in general. In fact, I am just an IT student right now, so I am not sure, if everything I am doing is the way you should do it. But it works, so you can look it up and improve it. I am going to put the whole project online, because I am currently not working at it anyways. There are still some bugs and stuff that needs to be fixed, but you should be able to find the solution for your problem.

**Comment by saroar** [comment]
@LisforLyance where are i find that code thanks for quick reply

**Comment by LisforLyance** [comment]
@saroar I am about to upload it right now, I am just not sure how to do it. Would you recommend using a repo or a gist for the whole project?

**Comment by saroar** [comment]
@LisforLyance you can put whole project if it is not ur private just create repo in github in ur account if you need any help we can call share ur screen :)

**Comment by LisforLyance** [comment]
@saroar I hope I did everything correct. You can find the repo here:
https://github.com/LisforLyance/ProjectFinder

If something doesn't work please let me know :)

**Comment by saroar** [comment]
thanks I will check sure

**Comment by saroar** [comment]
is this work with multiple users?

**Comment by saroar** [comment]
@LisforLyance you like Firebase i am using chat app with Swift Back end with real time chat app if you are interested we can work together with back api which will help you in future 
check this app https://github.com/saroar/RealtTimeChatWithAppPerfectWebSocketApi
real time chat like firebase but fully control by u :) send me ur email/fb if are interested more about it thanks

**Comment by LisforLyance** [comment]
@saroar The chat rooms in the app work with multiple users. The username labels are not loading the names jet, but I am sure you already figured that out. The problem should be described with comments. 

Your offer sounds nice and I will keep it in mind. I don't need it right now, as I'm just working at a small game app, but if I need it in the future I will talk to you :)

**Comment by LukA888** [comment]
Hi @AntonPalich! Did you finally write a tutorial about adding new items that behaves like messages (that support swipe to reveal feature, for example)?
